feat: support for deleting/resizing existing partitions

woodrowbarlow opened the pull request:
hi there @swysocki, great repo! i like your API design.

if you're curious: i'm evaluating this library for use in a test framework for embedded devices. i'll be taking disk images as input, making some changes to the partition layout, and then booting the modified images.

i'm making an attempt at supporting partition deletion/resizing. this would address issues #44 and #45.

this is still a completely untested work-in-progress. can you guide me on how to install dependencies and invoke the test suite?

swysocki commented:
@woodrowbarlow thank you for your interest!  Setup for development should be simple.  The following will get you started:

```bash
python3 -m venv .venv
source .venv/bin/activate
pip install -r requirements-dev.txt
```
Once that is done you can run the tests with
```bash
pytest tests/
```
If you are on a Linux machine you can install `sfdisk` and `sgdisk` to run `tests/test_integration.py`

woodrowbarlow commented:
thank you! i'll be back soon with new test cases. :)

woodrowbarlow commented:
p.s. i peeked your profile, saw you're in Raleigh and involved with Ampere. small world! i'm in Durham working with Analog Devices.

woodrowbarlow commented:
i believe this is ready for review. it's passing all tests locally, including integration tests.

your issue #44 talks about being able to find partitions by name or GUID so i extended the `find` function to support matching on UUID (case-insensitive). i've extended the corresponding test case as well.

i've extended the partitions array to support `remove` and `resize` functions, which can look up a partition by name or GUID. both of these support relocating the data within the disk image when the modifications are committed (this is useful when using your library to modify images that already have data). the `resize` function will truncate data without warning.

the data relocation works by creating a tempfile and copying data from the old location in the original image to the new location in the tempfile -- at the end, the original image is overwritten with the tempfile.

to make this work i had to make a couple changes:

* the MBR is now re-written during every commit. i hope this is not a problem.
* the LBA and size properties of the Partition are now `StagedAttribute`s -- you can write a value, but reading it back will continue to show the same value until the attributes are committed. there is a read-only property to get the staged value.

i've written new test cases for `remove` and `resize`.

your issue #45 talks about changing partition name and partition type. i haven't implemented anything to that effect; but is it not possible to simply modify those fields on the partition instance and then commit?

woodrowbarlow commented on gpt_image/partition.py:
```diff
@@ -175,46 +240,98 @@ def marshal(self) -> bytes:
+            f = tmpfile
+        try:
+            max_size = min(self.size, self.size_staged)
+            data = self.read(disk, max_size=max_size)
```
this section would probably benefit from some chunking (to avoid loading the entire partition into memory at once)

swysocki replied:
Agreed. This isn't a big issue for my use-case (the images are very small), so it can be addressed in a follow up.

woodrowbarlow commented:
how can i build this into a local package to import into other codebases, for further testing?

swysocki commented:
Thanks for the PR. I will give it a review as soon as I get a bit of time this week.

swysocki commented:
> how can i build this into a local package to import into other codebases, for further testing?

You can build a wheel locally with `python3 -m build --wheel`. This will allow you to import the `gpt_image` package in other projects.

swysocki commented:
@woodrowbarlow I think the changes are great. There are some linting (flake8, mypy) issues that need to be addressed, but other than that, this should be an easy update.

woodrowbarlow commented:
i've fixed the flake8 and mypy issues.

i also added "remove" and "resize" calls in the integration test and this revealed a bug. it looks like the primary and secondary headers weren't being properly un-marshalled when Disk.open is used. i've fixed that in a separate commit.

swysocki commented:
I merged your PR! Thanks again for the contribution. It looks like there was an issue with PYPI when the CI job ran and it didn't upload the package. I will have to manually push that now.  Until then you get the package from the Releases page.

woodrowbarlow commented:
yay! it was a pleasure to work with your code. :) thanks again!
